_test/test_instrument_classes/test_IX_detector_array_1.m

```diff
@@ -0,0 +1,554 @@
+            % Make from multiple detector banks, with different detector
+            % types
+            banks_123 = [obj.bank1, obj.bank2, obj.bank3];
+            D = IX_detector_array (banks_123, obj.bank4);
+            
+            banks_1234 = [obj.bank1 obj.bank2, obj.bank3, obj.bank4];
+            assertEqual (D.det_bank, banks_1234(:))
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_constructor_fromDataItems (obj)
+            % Check constructs with arguments as for IX_detector_bank
+            B = obj.bank1;
+            D = IX_detector_array (B);
+            Dtest = IX_detector_array (B.id, B.x2, B.phi,...
+                'det', B.det, 'rotvec', B.rotvec,'azim', B.azim);
+            assertEqual (D, Dtest)
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_constructor_fromDetparStruct (~)
+            id = (1:99)'; 
+            x2 = (0.0001:0.0001:0.0099)';
+            phi = (1.8:1.8:179.999)';
+            azim = (0.9:0.9:89.999)';
+            width = (0.02:0.0001:0.0298999)';
+            height = (0.03:0.0001:0.0398999)';
+            filename = 'random_name.par';
+            filepath = '';
+            
+            % Make detpar structure and use as input to constructor
+            detpar.group = id;
+            detpar.x2 = x2;
+            detpar.phi = phi;
+            detpar.azim = azim;
+            detpar.width = width;
+            detpar.height = height;
+            detpar.filepath = filepath;
+            detpar.filename = filename;
+            
+            D = IX_detector_array (detpar);
+            
+            % Construct directly
+            dets = IX_det_TobyfitClassic (width, height);
+            Dref = IX_detector_array (id, x2, phi, azim, dets);
+            Dref.filename = 'random_name.par';
+
+            assertEqual(D, Dref);
+        end
+        
+        %--------------------------------------------------------------------------
+        function test_constructor_4 (obj)
+            % Check fails if there are non-unique detector id values
+            bank2_tmp = obj.bank2;
+            bank2_tmp.id(4) = obj.bank1.id(2);
+            banks_123 = [obj.bank1, bank2_tmp, obj.bank3];
+            ME = assertExceptionThrown( ...
+                @()IX_detector_array (banks_123, obj.bank4), ...
+                'HERBERT:IX_detector_array:invalid_argument');
+            assertEqual(ME.message, 'Detector identifiers must all be unique')
+        end
+        
+        %--------------------------------------------------------------------------
+        %   Test changing properties
+        %--------------------------------------------------------------------------
+        function test_change_id (obj)
+            D = IX_detector_array (obj.bank1, obj.bank2, obj.bank3);
+            D.id = 101:113;
```

Comment:
There are two points here. 
(1) Unless Alex has done something I am unaware of, Horace as currently implemented assumes the detectors as identified internally are 1:ndet_tot; that is, the detector ids are currently not actually used. The ids are really just the names of the detectors, that just happen to be unique integers. Changing the ids therefore doesn't affect Horace. In the future it might be nice to have methods that actually refer to the ids, for example to filter out particular named detectors (i.e. those with a particular id). That would be definitely post-construction.
(2) However, notwithstanding the above, I think there is a point of principle. The functionality of how one might manipulate an object should not be dependent on how that ability might be abused in an application of the class. That should be a matter for protection in the particular software application that uses the class. Otherwise other use of the class has a restriction imposed by something completely unrelated. In this case, the API to an sqw object should require that changing the id updates both the detector ids and the pixels, just as changing the incident energy ei for an sqw object ought to update it in both the fermi chopper and the moderator. The reductio ad absurdum is that otherwise we should not be allowed to change ei for the chopper after construction. Would that then apply to not having setters for any property of any object?...